pkg/util/cache/lru.go

```diff
@@ -0,0 +1,216 @@
+	Put(k K, v V) error
+	// Unpin removes the pin from the key in the cache.
+	Unpin(k K)
+	// Evict removes one used key-value pair from the cache and returns the evicted key-value pair.
+	Evict() (K, V, error)
+	// Capacity returns the capacity of the cache.
+	Capacity() uint64
+	// Size returns the size of the cache.
+	Size() uint64
+	// Remove removes the key-value pair from the cache and returns the removed key-value pair.
+	// If a key is pinned, Remove will return false.
+	// If a key is not found, Remove will do nothing and return true.
+	Remove(k K) bool
+}
+
+type refItem[K comparable] struct {
+	ref uint
+	k   K
+}
+
+// LRUEvictor is an implementation of PinnableEvictor using LRU eviction.
+type LRUEvictor[K comparable] struct {
+	list *list.List // back is the least recently used
+	emap map[K]*list.Element
+}
+
+func (e *LRUEvictor[K]) Evict() (K, error) {
+	for i := e.list.Back(); i != nil; i = i.Prev() {
+		k := i.Value.(*refItem[K])
+		if k.ref > 0 {
+			continue
+		}
+		if _, ok := e.emap[k.k]; ok {
+			e.list.Remove(i)
+			delete(e.emap, k.k)
+			return k.k, nil
+		}
+	}
+
+	var r K
+	return r, ErrNoEvictionPossible
+}
+
+func (e *LRUEvictor[K]) Pin(k K) {
+	if v, ok := e.emap[k]; ok {
+		e.list.MoveToFront(v)
+		v.Value.(*refItem[K]).ref++
+		return
+	}
+	e.emap[k] = e.list.PushFront(&refItem[K]{k: k, ref: 1})
+}
+
+func (e *LRUEvictor[K]) Unpin(k K) {
+	if v, ok := e.emap[k]; ok {
+		v.Value.(*refItem[K]).ref--
+	}
+}
+
+func (e *LRUEvictor[K]) Remove(k K) bool {
+	if v, ok := e.emap[k]; ok {
+		if v.Value.(*refItem[K]).ref > 0 {
+			return false
+		}
+		e.list.Remove(v)
+		delete(e.emap, k)
+	}
+	return true
+}
+
+func NewLRUEvictor[K comparable]() *LRUEvictor[K] {
+	return &LRUEvictor[K]{list: list.New(), emap: make(map[K]*list.Element)}
+}
+
+type CacheImpl[K comparable, V any, E PinnableEvictor[K], M Measure[K, V]] struct {
+	capacity uint64
+	items    map[K]V
+
+	evictor E
+	measure M
+}
+
+func (c *CacheImpl[K, V, E, M]) Get(k K) (V, bool) {
+	if v, ok := c.items[k]; ok {
+		c.evictor.Pin(k)
+		return v, true
+	}
+
+	var r V
+	return r, false
+}
+
+func (c *CacheImpl[K, V, E, M]) Put(k K, v V) error {
+	if _, ok := c.items[k]; ok {
+		return ErrKeyAlreadyExists
+	}
+
+	c.measure.add(c.measure.measure(k, v))
+	c.items[k] = v
+	c.evictor.Pin(k)
+
+	for c.measure.get() > c.capacity {
+		k, err := c.evictor.Evict()
+		if err != nil {
+			return err
+		}
+
+		c.measure.sub(c.measure.measure(k, c.items[k]))
+		delete(c.items, k)
+	}
+
+	return nil
+}
+
+func (c *CacheImpl[K, V, E, M]) Unpin(k K) {
+	c.evictor.Unpin(k)
+}
+
+func (c *CacheImpl[K, V, E, M]) Evict() (K, V, error) {
+	var k K
+	var v V
+	var err error
+	k, err = c.evictor.Evict()
+	if err != nil {
+		return k, v, err
+	}
+
+	v = c.items[k]
+	c.measure.sub(c.measure.measure(k, v))
+	delete(c.items, k)
+
+	return k, v, nil
+}
+
+func (c *CacheImpl[K, V, E, M]) Capacity() uint64 {
+	return c.capacity
+}
+
+func (c *CacheImpl[K, V, E, M]) Size() uint64 {
```

Comment:
nit: better to rename `Size` as `Cost` and `measure` as `weight`, according to conventions of other cache implementations.